### ISSUE TITLE
Make ConstEmbeddingsGlove not use a ParameterCollection

### ISSUE THREAD

**Issue by MihaiSurdeanu** [clulab/processors#455]
The `ConstEmbeddingsGlove` class should not use a `ParameterCollection` but rather generate dynet embeddings on the fly directly from our compact representation.

**Comment by MihaiSurdeanu** [comment]
Hey @kwalcock:
Please look at the `ConstEmbeddingsGlove` implementation in the `constembed` branch:
https://github.com/clulab/processors/blob/constembed/main/src/main/scala/org/clulab/dynet/ConstEmbeddingsGlove.scala

The code seems to train and test with that. The NER performance is a bit lower, but that may be just randomization, maybe.

**Comment by kwalcock** [comment]
You noticed that the CompactWordEmbeddingMap does not have a getOrElse.  The WordEmbeddingMap has one, but it is implicit in getEmbedding.  There is a silent fallback there.  I should improve the consistency during the next round of revisions.  Also I wish that UNK was not a text that could conceivably be present in a document.  In other implementations, the empty word "" is used for this instead, because we're not allowing them in text.

**Comment by MihaiSurdeanu** [comment]
- I think we need to add the methods mentioned in the TODOs to CompactWordEmbeddingMap.
- I think we can use "" as the value for UNK. Maybe make it part of the CompactWordEmbeddingMap object, so it is consistently used across.
- Should I work on a CompactWordEmbeddingMap pool class that reuses one if already loaded? We can now do this.

**Comment by kwalcock** [comment]
It's difficult to imagine that you have time for it, but sure.  I'm not in the middle of it (yet).

Should it be able to load on its own the serialized version of the vectors?  It would be nice if those could be stored in artifactory and used when available.  It may be that the jar for that should not be zipped so that loading doesn't involve unzipping.  The space/time tradeoff may need to be evaluated.  I don't immediately see that processors includes any app that can be used to do the conversion from text to binary.  It's just a couple of lines to run loadTxt and then call save on the object.  It seems like the pool might need to use weak references, because there would otherwise be no way to release the vectors after something was all done with them.  Perhaps that should never happen anyway.  Alternatively, the pool could support some kind of close operation on vectors, maybe with an open count, so that the vectors could be GC'd when nobody wants them anymore.

I'm not sure about the values in the unknown vector.  Who decides it and when?  I was hoping to be able to replace any that is in the file by one coming from a client, perhaps in the constructor or factory object.  Perhaps it would only be needed in a getOrElse.  The get() could use the UNK value from the file, but the getOrElse(UNK) would take the one provided.  Whether or not there is one in the file depends on the program that wrote the binary file.  If the compact version is read from a text file instead, there normally isn't one at all.

These are incomplete thoughts hurried out in case you are already half finished.

**Comment by MihaiSurdeanu** [comment]
Good points. Let me see what I can do later today (probably in the evening). And then I'll pass it to you.

**Comment by kwalcock** [comment]
This strategy nicely shortens the code in ConstEmbeddingsGlove by 40 lines or so.  One of the lines that gets commented out is
```scala
val wordLookupParameters = parameters.addLookupParameters(w2i.size, Dim(dim))
```
The comment in ConstEmbeddings says "The generated Expressions must be const".  The C++ code has two kinds of options for lookup parameters:
```cpp
VariableIndex ComputationGraph::add_lookup(LookupParameter p, unsigned index);
VariableIndex ComputationGraph::add_const_lookup(LookupParameter p, unsigned index);
```
Are we sure it was done right previously?  Those lookup parameters aren't getting trained, are they?

**Comment by MihaiSurdeanu** [comment]
Not sure. I'd have to check.
But it doesn't matter anymore. We will skip this whole LookupParameter deal and create Expressions directly from our own map.

**Comment by kwalcock** [comment]
I suspect that the new version uses less memory at the same time because the it doesn't need to send the entire lookup vector, all of glove, to dynet,  If I understand correctly, it is going to be constantly sending 300-item arrays over as part of the expressions.  The data transfer itself would slow things down.  Also, the compact representation is not ideal for this.  It stores one humongous array and then slices it on demand.  The slicing takes time plus it means that sometimes two copies of the floats are being kept in Scala, plus the copy that probably goes to C++.  The non-compact version kept separate vectors, one for each line, so it could just return those directly.  (Previously we were multiplying lots of vectors and needing to created a new vectors to contain the result, so it was different and not necessary to keep separate vectors.)  It might eventually be advantageous to start caching some of the vectors that are being returned, vectors for stopwords, instead of constantly extracting them.

**Comment by MihaiSurdeanu** [comment]
Yeah... Should we use then the old WordEmbeddingMap class? I agree that this should reduce the number of array copies. 
If so, can you please add a WordEmbeddingMap.loadText that figures out if it needs to load from a file or a resource?

**Comment by MihaiSurdeanu** [comment]
We can both work in the `constembed` branch.

**Comment by kwalcock** [comment]
Re: // TODO: we need the length method implemented in CompactWordEmbeddingMap

+1.  In addition, it would be great to be able to process the text representation of glove without resorting to the line we have to add at the beginning to support.
```scala
val (wordCount, columns) = ...
```
The files could then be downloaded directly from Stanford without need for change.  In the compact version the number of words and floats are counted and the dimension is the quotient, so that part is taken care of if only the result could be queried.

**Comment by kwalcock** [comment]
Re: Expression.input(Dim(vector.length), new FloatVector(vector))

I don't know the details of this thing (yet), but for the record there is in C++ code but not in Scala, similar to the case with lookup parameters, this:
```cpp
VariableIndex ComputationGraph::add_const_parameters(Parameter p)
```
Is there any significant difference between an input expression and a const parameter?

**Comment by MihaiSurdeanu** [comment]
I don't think so. They should be roughly equivalent. 
My point here is to skip completely the ParameterCollection instance, which essentially replicates the functionality of our own embedding map class.

**Comment by MihaiSurdeanu** [comment]
I started the skeleton for the new embedding maps and pool in the `constembed` branch. In particular:
- Maps should implement this trait:
https://github.com/clulab/processors/blob/constembed/main/src/main/scala/org/clulab/embeddings/WordEmbeddingMap.scala

- The pool stores a collection of maps:
https://github.com/clulab/processors/blob/constembed/main/src/main/scala/org/clulab/embeddings/WordEmbeddingMapPool.scala

- I started a new ExplicitWordEmbeddingMap class, because the old one uses some old stuff (e.g., a ton of methods depend on sanitizeWord) that I would not change for now, because some legacy code depends on that:
https://github.com/clulab/processors/blob/constembed/main/src/main/scala/org/clulab/embeddings/ExplicitWordEmbeddingMap.scala

What do you think? The right direction?

**Comment by MihaiSurdeanu** [comment]
@kwalcock: I have a functioning implementation of this whole map pool thing in ConstEmbeddingsGlove. See:

- The pool class:
https://github.com/clulab/processors/blob/constembed/main/src/main/scala/org/clulab/embeddings/WordEmbeddingMapPool.scala

- And how it's used in here:
https://github.com/clulab/processors/blob/constembed/main/src/main/scala/org/clulab/dynet/ConstEmbeddingsGlove.scala

This is currently supported just for your Compact implementation.
Can you please take a look at this code? If everything seems legit, we can add the missing functionality. 
Thanks!

**Comment by MihaiSurdeanu** [comment]
Of course, the same pool can and should be used by Eidos, so we do not load embeddings more than once.

**Comment by kwalcock** [comment]
That's great.  I will look soon.  I'm about 2/3 through the LexiconNER case sensitivity.

**Comment by kwalcock** [comment]
FWIW I'm making changes and hopefully not duplicating effort.

If this is being revisited, it might be good to make explicit the fact that some arrays/vectors cannot be normalized by returning an Option and forcing the client code to deal with it.  Some unnormalized (zero) vectors are probably slipping by unnoticed and leading to wrong results.

I will at least cache the unknown vector which is being looked up time and time again.  Probably I will want to remove it even from the map.  Some versions of the vector file were created with the first vector being the unknown vector.  To know whether this was the unknown vector or some normal vector, because the reader does not know the details of the file, the word happened to be the special empty string.  If there wasn't an empty string, the vector file has no value for unknown.

I think the key used for the WordEmbeddingMapPool should not be a string but
```scala
case class Key(filename: String, compact: Boolean)
```

Perhaps the searching for the data should be done similarly to how fatdynet works.  It tries to find a file first, in case the user has overridden the resource.  This makes the resource argument unnecessary.  Whether it is cached (i.e., stored in binary format) or not could be determined by convention.  If they ask for org/clulab/glove/glove.840B.300d, a check is first made to see if it is in the pool.  Next, a check is made to find the file or the resource for org/clulab/glove/glove.840B.300d.bin, the faster-loading binary version.  Then if necessary it resorts to the file or resource for org/clulab/glove/glove.840B.300d.txt, the slower loading text version.  This would mean
```scala
def getOrElseCreate(filename: String, compact: Boolean = false)
```

More to follow...

**Comment by MihaiSurdeanu** [comment]
Thanks @kwalcock! A few comments:
- Why would an array not be possible to normalize? You mean when it's a zero vector? In such cases, by convention, the result is the same zero vector. I don't think we need an Option for this.
- Do you think caching the unknown vector would save a lot of time compared to a map lookup? Not sure that's the case, but it's worth testing.
- Agree with the Key class.
- I like the protocol for data searching! It's clear and reducing the number of parameters, which is always good.

**Comment by MihaiSurdeanu** [comment]
Now that we know we can control the data structure for embeddings, I wonder if we can compress the floats into 16-bit floats to save memory? There is research that shows that performance is not affected in a significant way if one does that. 
Then we can convert them to DyNet FloatVectors on the fly. What do you think @kwalcock ?

**Comment by kwalcock** [comment]
It sounds worthy of an experiment.  There are some type definitions in the newer word2vec that might be useful.  Can it wait long enough for me to finish up the previous changes?  Is there a handy-dandy implementation of 16-bit floats?  The ranges will be smaller, so I wonder if there will be more underflow errors and how they are handled.  Conversion will cost time and might be another reason to cache some of the values in separate arrays.  Right now there isn't a distinction between getting a read-only embedding for a word and getting an embedding that can be updated, for instance by having another array added to it.  DyNet is wanting read-only values that would not need to be copied every time.

**Comment by MihaiSurdeanu** [comment]
I sent an example implementation over email.

**Comment by kwalcock** [comment]
The pool is in the testing phase along with the new ExplicitWordEmbeddings.

**Comment by MihaiSurdeanu** [comment]
Nice! Please let me know when they pass. I'd like to run a test for Metal as well.

**Comment by kwalcock** [comment]
The current code has been pushed to https://github.com/clulab/processors/tree/kwalcock/constembed.  For the PR, the commits will be squashed and made tidy.  However, with this code, the following measurements of load times were made from TestWordEmbeddingMap reading from a file.  I will also eventually test reading from a jar, and runtime is also important.

| Class | Format | Time (ms) | Time (ms) |
| --- | --- | --: | --: |
| ExplicitWordEmbeddingMap |  txt | 57604 | 53138 |
| CompactWordEmbeddingMap | bin | 4901 | 5387 |

There is a 10x difference.

**Comment by MihaiSurdeanu** [comment]
Thanks!

I will test the runtime of the two formats as well. It may be that the explicit one is much faster at runtime than the other... Maybe, adding caching to the compact format is the way to close the gap, if this happens.

**Comment by kwalcock** [comment]
It could be that instead of returning a vector to callers we could return a view on a vector.  This makes a slice of a vector look like an independent vector and would require less copying.  If the interface is changing, it might be good to revisit what exactly clients need.

**Comment by MihaiSurdeanu** [comment]
To answer your view of a vector question, this `FloatVector` c-tor would have to be changed to accept a view here:
https://github.com/clulab/processors/blob/constembed/main/src/main/scala/org/clulab/dynet/ConstEmbeddingsGlove.scala#L21

This should be easy, I think...

**Comment by MihaiSurdeanu** [comment]
Also,
I compared the runtimes of the compact vs. explicit maps for an actual system (training and testing a NER). Logs are attached. The good news is that, after loading, there is some difference between the two maps, but it's not large. I think if we switched to views for the compact map, the difference will become negligible.

[
[log_compact.txt](https://github.com/clulab/processors/files/6084893/log_compact.txt)
](url)

[
[log_explicit.txt](https://github.com/clulab/processors/files/6084896/log_explicit.txt)
](url)

**Comment by kwalcock** [comment]
It would also be possible to request the Explicit version for long-running tasks.  The first minute of use is executed much more often than the tenth minute.  It might even be possible to change implementations in the middle of a run if it looks worthwhile.  That might be too fancy.  Right now, USE_COMPACT is exposed as a static var and can be changed by some client who knows that the slower loading time is worth the faster runtime.

**Comment by kwalcock** [comment]
Comments about those array views will be made momentarily.  I just added code so that the initial line with the dimensions in the glove files is not necessary.  It is now optional.  We should be able to use raw glove files without having to count lines in advance.

**Comment by kwalcock** [comment]
Eventually what is now returned from the embedding as an Array and which might be inadvertently mutated could be returned as a Seq to avoid the possibility.  It is suggested at https://stackoverflow.com/questions/4825702/why-no-immutable-arrays-in-scala-standard-library.  On first glance, it looks like an Array.view is an acceptable Seq, so maybe that works better all around.  The FloatVector takes in the constructor a Seq[Float] as well, so it *looks* like it is possible to use a View without changing FatDynet.  I'll know soon.

These FloatVectors are very inefficient.  Even in Vector.scala it looks like a copy or possibly two are made:
```scala
  def this(values: Seq[Float] = Seq.empty) {
    this(new internal.FloatVector(values.map(float2Float).asJavaCollection))
  }
```
Then to make the internal.FloatVector, this happens
```.java
  public FloatVector(java.util.Collection<Float> values) {
     this(values.size());
     int i = 0;
     for (java.util.Iterator<Float> it = values.iterator(); it.hasNext(); i++) {
         Float value = it.next();
         this.set(i, value);
     }
  }
```
so that this.set is called and makes for every single float involved a trip into C++ and back.  I'll double check whether that is really necessary.  If I ever get to that Scala implementation, it won't be.

**Comment by kwalcock** [comment]
The view thing seems to have worked and is in the branch kwalcock/constembed (which could be merged into constembed).

`CompactWordEmbeddingMap.get()` was changed from
```scala
      array.slice(offset, offset + columns)
```
which makes a copy to
```scala
      array.view(offset, offset + columns)
```
which doesn't.  `ConstEmbeddingsGlove` didn't need to be changed yet.

I will try the 16-bit version soon.  At least for this code, the return types are OK.  I'm not yet sure for other clients like reach or eidos.

**Comment by MihaiSurdeanu** [comment]
Thanks @kwalcock ! I will test this later today.
A few questions/comments:
- Should we use Seq or IndexedSeq for these views? I have used IndexedSeq in my code to indicate that the data structures underneath are indexed.
- Ideally, we should reduce the number of unnecessary copies in FloatVector (and other vectors). For example, it may not be necessary to go through the intermediate java collection?

**Comment by MihaiSurdeanu** [comment]
The runtime does not seem to be impacted in any significant way by this change. This may indicate that there are bigger costs in other parts of dynet.

**Comment by MihaiSurdeanu** [comment]
I think we can prepare a PR to merge what we have in master? 
Then address the 16 bit as a separate issue?

**Comment by kwalcock** [comment]
I added code in the apply methods to catch unknown vectors and route them to a variable so they do not appear in the map at all.  I will publish locally and see what breaks.  I suspect that quite a lot will.  If the ValueType is changed from Float to java.lang.Float, then one of the conversions, float2Float, is unnecessary.  I'm not sure what the .asJavaCollection does to the Seq, but will find out.  It could be that this conversion is a no-op or can be avoided by using Java containers if necessary.

**Comment by MihaiSurdeanu** [comment]
Just for the record, I am adding here the log when using the compact map with view to train the same NER.
Note that currently the compact map yields a runtime that is 10 - 20% slower than the explicit map. So, caching might be the way to speed it up further. Possibly having a dedicated variable for UNK too...
[log_compact_withview.txt](https://github.com/clulab/processors/files/6095832/log_compact_withview.txt)

**Comment by kwalcock** [comment]
It's acting strange right now and it occurs to me that the strategy with the Pool has the disadvantage of synchronization.  In the past it was possible to load two vector files at the same time and it looks like that was happening.  Eidos startup is multithreaded and processors was started in parallel with the loading of the vectors for Eidos grounding.  The this.synchronized in WordEmbeddingMapPool.getOrElseUpdate would seem to prevent that (although I see evidence to the contrary that it is working as intended just now).  This would slow down startup time.  It may eventually be necessary to make use of Futures or a similar mechanism to overcome the bottleneck.

**Comment by MihaiSurdeanu** [comment]
Agreed. BUT, they should not be loading two different maps. Eidos should reuse the map loaded by processors. There should be a single load.

**Comment by kwalcock** [comment]
It felt like the load time was increasing after an array buffer was added so that the dimensions do not need to be specified.  It seems to be that case.  I'll get that fixed.  The next measurements will include some test of the time to convert into an ExpressionVector.

| time (ms) | useFileElseResource | useTxtElseBin | useExplicitElseCompact | useOldElseNew |
| --: | --- | --- | --- | --- |
|  48778 | true  | true  | true  | true  |
|  53176 | false | true  | true  | true  |
|  32539 | true  | true  | false | true  |
|  40544 | false | true  | false | true  |
|   2272 | true  | false | false | true  |
|  49776 | true  | true  | true  | false |
|  61946 | false | true  | true  | false |
|  25327 | true  | false | true  | false |
| 173111 | true  | true  | false | false |
| 142309 | false | true  | false | false |
|   4041 | true  | false | false | false |

**Comment by MihaiSurdeanu** [comment]
Kinda hard to parse this table... Which combination had 4041 load time?

**Comment by kwalcock** [comment]
File (rather than resource), binary (rather than text), compact (rather than explicit), new (rather than the old/previous version in master which had been faster, at least with only one sample).

**Comment by MihaiSurdeanu** [comment]
Thanks!

I think we need to optimize the loading from resource, which is the most common scenario. Although we can change that...

**Comment by kwalcock** [comment]
Definitely.  Some buffering recently added had a detrimental effect.  In the end I think we'll want to publish the binary version to artifactory as well.  Processors will transparently choose that one if it is available.

**Comment by MihaiSurdeanu** [comment]
Do you have a config with useFileElseResource = false, useTxtElseBin = false?

**Comment by kwalcock** [comment]
Not yet.  I haven't set up the (local) publishing of those files.  In general there is a penalty for the resources over the files, probably because the resources are compressed.  It is possible to make an uncompressed jar if necessary.

**Comment by kwalcock** [comment]
I also didn't yet figure out how the old version is stored in binary format.  There is code for readBinaryMatrix, but I didn't yet find the writeBinaryMatrix.  The nearby saveMatrix writes in text format.

**Comment by MihaiSurdeanu** [comment]
I think it was just standard Java serialization of the whole object.

**Comment by kwalcock** [comment]
Here is a more complete set of performance measurements.  I am working on improvements to the non-Old versions.

| operation | time (ms) | class | location | useFileElseResource | useTxtElseBin | useExplicitElseCompact | useOldElseNew |
| --------- | --------- | ----- | -------- | ------------------- | ------------- | ---------------------- | ------------- |
| load | 47631 | OldWordEmbeddingMap | ../glove.840B.300d.10f.txt | true | true | true | true |
| load | 54248 | OldWordEmbeddingMap | /org/clulab/glove/glove.840B.300d.10f.txt | false | true | true | true |
| load | 28166 | OldWordEmbeddingMap | ../glove.840B.300d.10f.bin.explicit.old | true | false | true | true |
| load | 32681 | OldCompactWordEmbeddingMap | ../glove.840B.300d.10f.txt | true | true | false | true |
| load | 40512 | OldCompactWordEmbeddingMap | /org/clulab/glove/glove.840B.300d.10f.txt | false | true | false | true |
| load | 3070 | OldCompactWordEmbeddingMap | ../glove.840B.300d.10f.bin.compact.old | true | false | false | true |
| load | 46461 | ExplicitWordEmbeddingMap | ../glove.840B.300d.10f.txt | true | true | true | false |
| load | 54962 | ExplicitWordEmbeddingMap | /org/clulab/glove/glove.840B.300d.10f.txt | false | true | true | false |
| load | 25604 | ExplicitWordEmbeddingMap | ../glove.840B.300d.10f.bin.explicit.new | true | false | true | false |
| load | 198374 | CompactWordEmbeddingMap | ../glove.840B.300d.10f.txt | true | true | false | false |
| load | 200831 | CompactWordEmbeddingMap | /org/clulab/glove/glove.840B.300d.10f.txt | false | true | false | false |
| load | 4259 | CompactWordEmbeddingMap | ../glove.840B.300d.10f.bin.compact.new | true | false | false | false |


| operation | time (ms) | class | location | useFileElseResource | useTxtElseBin | useExplicitElseCompact | useOldElseNew |
| --------- | --------- | ----- | -------- | ------------------- | ------------- | ---------------------- | ------------- |
| run | 9404 | OldWordEmbeddingMap | ../glove.840B.300d.10f.txt | true | true | true | true |
| run | 9763 | OldWordEmbeddingMap | /org/clulab/glove/glove.840B.300d.10f.txt | false | true | true | true |
| run | 10375 | OldWordEmbeddingMap | ../glove.840B.300d.10f.bin.explicit.old | true | false | true | true |
| run | 9307 | OldCompactWordEmbeddingMap | ../glove.840B.300d.10f.txt | true | true | false | true |
| run | 8626 | OldCompactWordEmbeddingMap | /org/clulab/glove/glove.840B.300d.10f.txt | false | true | false | true |
| run | 9296 | OldCompactWordEmbeddingMap | ../glove.840B.300d.10f.bin.compact.old | true | false | false | true |
| run | 9571 | ExplicitWordEmbeddingMap | ../glove.840B.300d.10f.txt | true | true | true | false |
| run | 9516 | ExplicitWordEmbeddingMap | /org/clulab/glove/glove.840B.300d.10f.txt | false | true | true | false |
| run | 10130 | ExplicitWordEmbeddingMap | ../glove.840B.300d.10f.bin.explicit.new | true | false | true | false |
| run | 11285 | CompactWordEmbeddingMap | ../glove.840B.300d.10f.txt | true | true | false | false |
| run | 11355 | CompactWordEmbeddingMap | /org/clulab/glove/glove.840B.300d.10f.txt | false | true | false | false |
| run | 9565 | CompactWordEmbeddingMap | ../glove.840B.300d.10f.bin.compact.new | true | false | false | false |

**Comment by MihaiSurdeanu** [comment]
It seems the ExplicitWordEmbedding wins over the CompactWordEmbedding for both load and run times.
I suspect the load times from resource can be improved significantly if we store the binaries.

Another important statistic would be: what is the memory footprint of Explicit vs. Compact? If not a big difference, we might as well go with the explicit representation...

**Comment by kwalcock** [comment]
That 200,000 value was cut to 50,000 last night and more will follow shortly.  Still, 50 seconds is too long to wait.  5 is more like it.

**Comment by kwalcock** [comment]
Those 200,000 values for loading are gone.  There's quite a bit of variation in individual values and probably some ordering effects.  Running time is next.

| operation | time (ms) | class | location | useFileElseResource | useTxtElseBin | useExplicitElseCompact | useOldElseNew |
| --------- | --------: | ----- | -------- | ------------------- | ------------- | ---------------------- | ------------- |
| load | 46637 | OldWordEmbeddingMap | ../glove.840B.300d.10f.txt | true | true | true | true |
| load | 52932 | OldWordEmbeddingMap | /org/clulab/glove/glove.840B.300d.10f.txt | false | true | true | true |
| load | 28956 | OldWordEmbeddingMap | ../glove.840B.300d.10f.bin.explicit.old | true | false | true | true |
| load | 29926 | OldCompactWordEmbeddingMap | ../glove.840B.300d.10f.txt | true | true | false | true |
| load | 38450 | OldCompactWordEmbeddingMap | /org/clulab/glove/glove.840B.300d.10f.txt | false | true | false | true |
| load | 3146 | OldCompactWordEmbeddingMap | ../glove.840B.300d.10f.bin.compact.old | true | false | false | true |
| load | 50693 | ExplicitWordEmbeddingMap | ../glove.840B.300d.10f.txt | true | true | true | false |
| load | 60958 | ExplicitWordEmbeddingMap | /org/clulab/glove/glove.840B.300d.10f.txt | false | true | true | false |
| load | 24432 | ExplicitWordEmbeddingMap | ../glove.840B.300d.10f.bin.explicit.new | true | false | true | false |
| load | 32502 | CompactWordEmbeddingMap | ../glove.840B.300d.10f.txt | true | true | false | false |
| load | 42987 | CompactWordEmbeddingMap | /org/clulab/glove/glove.840B.300d.10f.txt | false | true | false | false |
| load | 4262 | CompactWordEmbeddingMap | ../glove.840B.300d.10f.bin.compact.new | true | false | false | false |

| operation | time (ms) | class | location | useFileElseResource | useTxtElseBin | useExplicitElseCompact | useOldElseNew |
| --------- | --------: | ----- | -------- | ------------------- | ------------- | ---------------------- | ------------- |
| run | 7102 | OldWordEmbeddingMap | ../glove.840B.300d.10f.txt | true | true | true | true |
| run | 8058 | OldWordEmbeddingMap | /org/clulab/glove/glove.840B.300d.10f.txt | false | true | true | true |
| run | 7499 | OldWordEmbeddingMap | ../glove.840B.300d.10f.bin.explicit.old | true | false | true | true |
| run | 6837 | OldCompactWordEmbeddingMap | ../glove.840B.300d.10f.txt | true | true | false | true |
| run | 8307 | OldCompactWordEmbeddingMap | /org/clulab/glove/glove.840B.300d.10f.txt | false | true | false | true |
| run | 7414 | OldCompactWordEmbeddingMap | ../glove.840B.300d.10f.bin.compact.old | true | false | false | true |
| run | 10757 | ExplicitWordEmbeddingMap | ../glove.840B.300d.10f.txt | true | true | true | false |
| run | 11023 | ExplicitWordEmbeddingMap | /org/clulab/glove/glove.840B.300d.10f.txt | false | true | true | false |
| run | 11042 | ExplicitWordEmbeddingMap | ../glove.840B.300d.10f.bin.explicit.new | true | false | true | false |
| run | 7728 | CompactWordEmbeddingMap | ../glove.840B.300d.10f.txt | true | true | false | false |
| run | 8418 | CompactWordEmbeddingMap | /org/clulab/glove/glove.840B.300d.10f.txt | false | true | false | false |
| run | 9331 | CompactWordEmbeddingMap | ../glove.840B.300d.10f.bin.compact.new | true | false | false | false |

**Comment by kwalcock** [comment]
A relatively simple change to fatdynet saved about 30% on the runtime.  Each of the four groups of three are the same embedder, just fetched in different ways, and should really measure the same time if it weren't for things like garbage collection.  I'll put it through the profiler next.

| operation | time (ms) | class | location | useFileElseResource | useTxtElseBin | useExplicitElseCompact | useOldElseNew |
| --------- | --------- | ----- | -------- | ------------------- | ------------- | ---------------------- | ------------- |
| run | 4781 | OldWordEmbeddingMap | ../glove.840B.300d.10f.txt | true | true | true | true |
| run | 4804 | OldWordEmbeddingMap | /org/clulab/glove/glove.840B.300d.10f.txt | false | true | true | true |
| run | 7844 | OldWordEmbeddingMap | ../glove.840B.300d.10f.bin.explicit.old | true | false | true | true |
| run | 4566 | OldCompactWordEmbeddingMap | ../glove.840B.300d.10f.txt | true | true | false | true |
| run | 4989 | OldCompactWordEmbeddingMap | /org/clulab/glove/glove.840B.300d.10f.txt | false | true | false | true |
| run | 4727 | OldCompactWordEmbeddingMap | ../glove.840B.300d.10f.bin.compact.old | true | false | false | true |
| run | 6967 | ExplicitWordEmbeddingMap | ../glove.840B.300d.10f.txt | true | true | true | false |
| run | 7205 | ExplicitWordEmbeddingMap | /org/clulab/glove/glove.840B.300d.10f.txt | false | true | true | false |
| run | 7035 | ExplicitWordEmbeddingMap | ../glove.840B.300d.10f.bin.explicit.new | true | false | true | false |
| run | 5946 | CompactWordEmbeddingMap | ../glove.840B.300d.10f.txt | true | true | false | false |
| run | 5540 | CompactWordEmbeddingMap | /org/clulab/glove/glove.840B.300d.10f.txt | false | true | false | false |
| run | 7318 | CompactWordEmbeddingMap | ../glove.840B.300d.10f.bin.compact.new | true | false | false | false |

**Comment by MihaiSurdeanu** [comment]
Nice! What did you change?

**Comment by kwalcock** [comment]
This
```scala
  def this(values: Seq[Float] = Seq.empty) {
    this(new internal.FloatVector(values.map(float2Float).asJavaCollection))
  }
```
was changed to
```scala
  def addAll(values: Seq[Float]): Unit = {
    var i = 0 // optimization
    while (i < values.length) {
      vector.set(i, values(i))
      i += 1
    }
  }
```

**Comment by MihaiSurdeanu** [comment]
Nice. Any other vectors that should be changed in a similar way?

**Comment by kwalcock** [comment]
I will double check soon.  To use the above code, `ConstEmbeddingsGlove` was changed from
```scala
  override def mkEmbedding(word:String): Expression = {
    val vector = wordVectors.getOrElseUnknown(word)
    Expression.input(Dim(dim), new FloatVector(vector))
  }
```
to
```scala
  override def mkEmbedding(word: String): Expression = {
    val embeddingSeq: IndexedSeq[Float] = wordEmbeddingMap.getOrElseUnknown(word)
    val floatVector = new FloatVector(embeddingSeq.length)
    floatVector.addAll(embeddingSeq)
    Expression.input(dynetDim, floatVector)
  }
```
and I'll still make that a little prettier, possibly a utility method somewhere.

**Comment by kwalcock** [comment]
See #460 among others.
